### PR TITLE
Makefile

### PR DESCRIPTION
**Pull request by garloff**
This adds a Makefile that allows all kinds of shortcuts.
Helps my memory, saves typing and avoids mistakes of my poor old brain.

I added --timeout to the invocations. At least when doing it all (infra+ceph+openstack), one hour (the default heat timeout) is risky, so also add this to the example in the documentation.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+stack.yml: templates/stack.yml.j2
+	jinja2 -o $@ $(TMPL_PARAMS) $^
+
+stack-single.yml: templates/stack.yml.j2
+	jinja2 -o $@ $(TMPL_PARAMS) -Dnumber_of_nodes=0 $^
+
+dry-run: stack.yml environment.yml
+	openstack stack create --dry-run -t $< -e environment.yml $(STACK_PARAMS) $(STACKNAME) -f json; echo
+
+deploy: stack.yml environment.yml
```
Comment (berendt) [inline Makefile]: That's unclear. I would use create here instead of deploy.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+deploy-infra: stack.yml environment.yml
+	@touch .deploy.$(STACKNAME)
+	openstack stack create -t $< -e environment.yml $(STACK_PARAMS) --parameter deploy_infrastructure=true $(STACKNAME)
+
+deploy-infra-ceph: stack.yml environment.yml
+	@touch .deploy.$(STACKNAME)
+	openstack stack create -t $< -e environment.yml $(STACK_PARAMS) --parameter deploy_infrastructure=true --parameter deploy_ceph=true $(STACKNAME)
+
+deploy-infra-ceph-openstack: stack.yml environment.yml
```
Comment (berendt) [inline Makefile]: OpenStack depends on Infra and Ceph. I think "deploy-openstack" or "deploy" is the best choice.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+deploy: stack.yml environment.yml
+	@touch .deploy.$(STACKNAME)
+	openstack stack create -t $< -e environment.yml $(STACK_PARAMS) $(STACKNAME)
+
+deploy-infra: stack.yml environment.yml
+	@touch .deploy.$(STACKNAME)
+	openstack stack create -t $< -e environment.yml $(STACK_PARAMS) --parameter deploy_infrastructure=true $(STACKNAME)
+
+deploy-infra-ceph: stack.yml environment.yml
```
Comment (berendt) [inline Makefile]: I think it makes the most sense to have a deploy-ceph only. Ceph doesn't need Infra.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	@MANAGER_ADDRESS=$$(openstack stack output show $(STACKNAME) manager_address -f value -c output_value); \
+	echo "MANAGER_ADDRESS=$$MANAGER_ADDRESS" > $@; \
+	echo $$MANAGER_ADDRESS
+
+# Convenience: sshuttle invocation
+sshuttle: ~/.ssh/id_rsa.$(STACKNAME) .MANAGER_ADDRESS.$(STACKNAME)
+	source ./.MANAGER_ADDRESS.$(STACKNAME); \
+	sshuttle --ssh-cmd "ssh -i $<" -r dragon@$$MANAGER_ADDRESS 192.168.40.0/24 192.168.50.0/24 192.168.90.0/24
+
+ssh_manager: ~/.ssh/id_rsa.$(STACKNAME) .MANAGER_ADDRESS.$(STACKNAME)
```
Comment (berendt) [inline Makefile]: I think ssh makes more sense here. You can only connect to the manager via SSH.

Comment (berendt) [reply Makefile]:
```
STAT=$(openstack stack list -f value -c "Stack Name" -c "Stack Status" | grep testbed | cut -d' ' -f2); \
	if test -n "$STAT"; then touch .deploy.testbed; else echo "use make deploy or deploy-infra or ...."; exit 1; fi
Missing value auth-url required for auth plugin password
use make deploy or deploy-infra or ....
make: *** [.deploy.testbed] Error 1
```

Comment (berendt) [reply Makefile]: ```export OS_CLOUD=testbed``` required. 

I think it makes sense to introduce --os-cloud $OS_CLOUD and set OS_CLOUD to testbed by default.

File: Makefile (hunk 1 above)
Comment (garloff) [reply Makefile]: If we have targets like deploy-infra, deploy-ceph, deploy-openstack, isn't a nakes deploy the most logical. Happy to create a "create" alias, though, if that works for you.

File: Makefile (hunk 2 above)
Comment (garloff) [reply Makefile]: deploy-openstack makes sense to me.

File: Makefile (hunk 3 above)
Comment (garloff) [reply Makefile]: OK, will change to deploy-ceph and leave out infra.

File: Makefile (hunk 4 above)
Comment (garloff) [reply Makefile]: make ssh is fine with me.
I don't want to force the usage of --os-cloud, so I'll have to do a bit of magic to not break parameterless openstack invocation either (both classic OS_ style config or exported OS_CLOUD var).

Comment (garloff) [reply Makefile]: OK, so all cases work now.
1. Old config (OS_PASSWORD)
2. New config with OS_CLOUD exported
3. New config without OS_CLOUD, passing --os-cloud=testbed
4. New config without OS_CLOUD, make param OS_CLOUD=override, passing --os-cloud=override.

File: README.md [hunk 5]
```diff
@@ -41,9 +41,9 @@ The following stable releases are supported. The development branch usually work
 ## Test status of cloud providers
 
 * [Betacloud](https://www.betacloud.de): Works
-* [Citycloud](https://www.citycloud.com): Works (need to change disk names from sdX to vdX before deploying Ceph on manager node ``/opt/configuration/inventory/host_vars/testbed-node-?.osism.local.yml``)
-* [OTC](https://open-telekom-cloud.com/): Needs ``enable_snat``, ``enable_dhcp`` and older heat, still fails
-* [teuto.stack](https://teutostack.de/): Currently lacks support for heat
+* [Citycloud](https://www.citycloud.com): Works (need to change disk names from sdX to vdX, pass ``drives_vdx: true`` in environment)
```
Comment (berendt) [inline README.md]: This is not part of this PR and is already merged. Please do a manual rebase.

File: README.md [hunk 6]
```diff
@@ -41,9 +41,9 @@ The following stable releases are supported. The development branch usually work
...
+* [OTC](https://open-telekom-cloud.com/): Needs ``enable_snat``, ``enable_dhcp``, ``dns_nameservers``, fixup NIC names via custom runcmd commands and an older heatversion. It also needs two cloud-init patches to get get userdata.
```
Comment (berendt) [inline README.md]: Dito.

File: README.md [hunk 7]
```diff
@@ -41,9 +41,9 @@ The following stable releases are supported. The development branch usually work
...
+* [teuto.stack](https://teutostack.de/): Currently lacks support for heat.
```
Comment (berendt) [inline README.md]: Dito.

File: README.md [hunk 8]
```diff
@@ -419,9 +442,19 @@ openstack --os-cloud testbed \
...
+  --timeout 9000 \
   -t stack.yml testbed
 ```
 
+The ``--timeout 9000`` parameter avoids heat giving up too early.
+(The default timeout for heat stacks is typically 3600.)
+
+This can also be achieved using ``make deploy-openstack``.
+
+The parameter ``--parameter drives_vdx=true`` can be passed (or ``drives_vdx: true`` be set
```
Comment (berendt) [inline README.md]: Dito.

File: README.md (hunk 5 above)
Comment (berendt) [reply README.md]: Ignore this...

**Comment by garloff** [comment]
There was a rebase lacking ... but you resolved it all yourself :-) THANKS!
